cmd/osbuild-worker/main.go

```diff
@@ -459,38 +450,21 @@ func main() {
 
 		fmt.Printf("Running job %v\n", job.Id())
 
-		ctx, cancel := context.WithCancel(context.Background())
+		ctx, cancelWatcher := context.WithCancel(context.Background())
 		go WatchJob(ctx, job)
 
 		result, err := RunJob(job, store, kojiServers)
-		if err != nil || result.Success == false {
-			log.Printf("  Job failed: %v", err)
-
-			// Ensure we always have a non-nil result, composer doesn't like nils.
-			// This can happen in cases when OSBuild crashes and doesn't produce
-			// a meaningful output. E.g. when the machine runs of disk space.
-			if result == nil {
-				result = &osbuild.Result{
-					Success: false,
-				}
-			}
-
-			// set the success to false on every error. This is hacky but composer
-			// currently relies only on this flag to decide whether a compose was
-			// successful. There's no different way how to inform composer that
-			// e.g. an upload fail. Therefore, this line reuses the osbuild success
-			// flag to indicate all error kinds.
-			result.Success = false
-		} else {
-			log.Printf("  🎉 Job completed successfully: %v", job.Id())
+		cancelWatcher()
+		if err != nil {
+			log.Printf("Job %s failed: %v", job.Id(), err)
+			continue
```

Comment:
I'm afraid this change will lead to a job stuck forever in the `WAITING` state. We must somehow send information to `composer` that a job failed even in cases where it wasn't an osbuild error.

Reply:
Good catch! This set of patches sends more errors in a structured way (i.e., the ones from targets), but doesn't set `osbuild.Result.Success` anymore on other errors. This was the wrong thing to do, but at least the job was marked as failed in composer.

The remaining possible errors are related to the worker:
1. not being able to create temporary directories
2. OSBuild returning invalid JSON
3. not being able to talk to composer.

We should not try to communicate (3) to composer, because the error itself was not being able to reach it.

The only thing composer can do for errors (1) and (2) (and possibly others in that category) is to reschedule the job on a different worker. That's something composer will need to do very soon anyway, when a job times out or a watchdog notices a worker is gone. For example, because we're using spot instances on AWS. If we don't add it, misbehaving workers can swallow all jobs, which will then never be completed.

I hesitate to add this logic with this PR, because it's already quite big. But I also hesitate to add another way of reporting errors of the (1) and (2) type, given that we want this other solution soon anyway.

Would you be ok merging this without, with this (rather small?) regression – that some jobs which used to be marked as failed, with no error message, are now marked as pending. In either case, someone with access to the worker has to check its logs.

Reply:
> Good catch! This set of patches sends more errors in a structured way (i.e., the ones from targets), but doesn't set `osbuild.Result.Success` anymore on other errors. This was the wrong thing to do, but at least the job was marked as failed in composer.
> 
> The remaining possible errors are related to the worker:
> 
>     1. not being able to create temporary directories
> 
>     2. OSBuild returning invalid JSON
> 
>     3. not being able to talk to composer.
> 
> 
> We should not try to communicate (3) to composer, because the error itself was not being able to reach it.
> 
> The only thing composer can do for errors (1) and (2) (and possibly others in that category) is to reschedule the job on a different worker.

This is the part I was missing. These errors are indeed caused by a misbehaving worker. I think that the worker should return some kind of "misbehaved" status at some point so we don't rely solely on timeouts.

I'm OK to merge with this small regression, I'm glad we cleared things up. Thanks!